pkg/provisioning/import.go

```diff
@@ -0,0 +1,712 @@
+// Copyright © 2023 Meroxa, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package provisioning
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/conduitio/conduit/pkg/connector"
+	"github.com/conduitio/conduit/pkg/foundation/cerrors"
+	"github.com/conduitio/conduit/pkg/pipeline"
+	"github.com/conduitio/conduit/pkg/processor"
+	"github.com/conduitio/conduit/pkg/provisioning/config"
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+)
+
+// Import takes a pipeline config and imports it into Conduit.
+func (s *Service) Import(ctx context.Context, newConfig config.Pipeline) error {
+	oldConfig, err := s.Export(ctx, newConfig.ID)
+	if err != nil && !cerrors.Is(err, pipeline.ErrInstanceNotFound) {
+		s.logger.Warn(ctx).Err(err).Msgf("could not export pipeline with ID %v, trying to provision new pipeline despite the error", newConfig.ID)
+		// TODO does this behavior make sense? Should we rather return the error
+		//  and skip provisioning?
```

Comment:
If there's an error, `oldConfig` should be an empty struct, so we're overwriting it, whatever it may be. `oldConfig` being empty means that lifecycle events would see an incorrect previous state, so cleanups may not happen.

Reply:
First of all, let's see when can we land in this condition - the only errors that `Export` can encounter are `*.ErrInstanceNotFound`. Now, if the pipeline is not found that's fine, in that case we just provision the new pipeline. But if a connector or processor is not found that's strange, it would mean that the pipeline references a connector or processor that doesn't exist. Essentially we would need to have a corrupted pipeline in Conduit to begin with.

You are right that we will overwrite the entities if they exist, because the pipeline/connector/processor `Service` doesn't check if the ID is already taken (this is done in the orchestrator and here in `Import` when exporting the existing pipeline). So we should be able to provision the new pipeline, _but_ there might be some old connectors or processors that are left in Conduit but not connected to the pipeline (the new pipeline won't point to them). I don't think they will interfere with the newly provisioned pipeline though, i.e. the new pipeline should be able to start and stop perfectly fine.

Connector states are not preserved in this case because connectors are _created_ and not _updated_.

To sum up: this only happens if the state is corrupted in the first place (pipeline references connectors/processors that don't exist). If the config is valid the pipeline should be provisioned correctly but there might be some corrupted state left over (i.e. dangling connectors and/or processors). Connector states are discarded as well.

Reply:
Btw, it's good you opened the discussion, that's exactly what I expected when I wrote that comment. If we agree that the behavior is what we want, I'll remove the comment and add an explanation what can happen in that case.

Reply:
> First of all, let's see when can we land in this condition - the only errors that `Export` can encounter are `*.ErrInstanceNotFound` (`*` can be `pipeline`, `connector` or `processor`). Now, if the pipeline is not found that's fine, in that case we just provision the new pipeline. But if a connector or processor is not found that's strange, it would mean that the pipeline references a connector or processor that doesn't exist. Essentially we would need to have a corrupted pipeline in Conduit to begin with.

Thanks for this! My thinking is that: at this point we know and can't see any other errors besides `ABCNotFound`. That may change in future, at which point we might get unexpected results. I also can't think of any other error being returned besides `ABCNotFound`, so I think it would be good to add that to the method comment (i.e. that the method returns a nil error or `NotFound`).

Reply:
We're on the same page!